### DISCUSSION TITLE
Help writing FCPXML Documentation?

### DISCUSSION THREAD

**Discussion by latenitefilms** [orchetect/swift-daw-file-tools#27] [Q&A]
Even though I've developed lots of FCPXML-based applications over the years, I still find it insanely confusing and I constantly forget how to do things and how things work.

To combat this, I started writing some documentation on FCP Cafe to determine the "source timecode" of a `sync-clip`.

However, I'm having an insanely hard time writing something that actually works.

Here's what I've got thus far:

https://fcp.cafe/developers/fcpxml/#getting-source-timecode

Even ChatGPT struggles to understand how to get the right answer.

Can anyone help me solve this? It SHOULD be such a simple example, but alas, even I can't get my head around it.

I wanted to write this "simple" one first, before even attempting to talk about different frame rates.

Any ideas?

**Comment by orchetect** [comment]
I wrote an algorithm that can calculate the 'absolute timecode' of any element in DAWFileKit and it works with everything I've thrown at it so far. (There may be some uncovered edge cases where it needs work yet.)

I'm not sure just reading the source code will give you what you want however because there are some helper methods that may be tricky to work out. There were numerous special cases and handlers for specific hierarchies because of how nuanced FCPXML can be.

**Reply by latenitefilms** [reply]
Ironically, I've solved it many times before in code... it's writing it out in "plain" English that I'm struggling with in this very specific example.

**Reply by orchetect** [reply]
Actually yes - I tried a few attempts at writing a plain-English algorithm or some type of flow chart and scrapped it early on as it actually only made things more confusing for myself 😆

**Comment by latenitefilms** [comment]
After a lot of swearing, I THINK I have something that finally makes sense:

https://fcp.cafe/developers/fcpxml/#getting-source-timecode
